### PR TITLE
fix warmup_bias_lr default for MuSGD optimizer

### LINKED ISSUE #23637: Title: optimizer=auto incorrectly sets warmup_bias_lr=0.0 when MuSGD is selected
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

_No response_

### Bug

When optimizer=auto resolves to MuSGD (iterations > 10000), warmup_bias_lr is incorrectly forced to 0.0. This setting is intended for Adam-family optimizers only, but
  currently applies unconditionally to both AdamW and MuSGD.

  Location

  ultralytics/engine/trainer.py, build_optimizer() method:

  # Line 951-952
  name, lr, momentum = ("MuSGD", 0.01, 0.9) if iterations > 10000 else ("AdamW", lr_fit, 0.9)
  self.args.warmup_bias_lr = 0.0  # no higher than 0.01 for Adam  <-- always executes

  The comment says "no higher than 0.01 for Adam", but this line executes regardless of whether MuSGD or AdamW is selected.

  Impact

  - SGD/MuSGD benefits from a non-zero warmup_bias_lr (default 0.1) during warmup to stabilize early training.
  - With warmup_bias_lr=0.0, bias parameters receive zero learning rate at the start of warmup, which can degrade final accuracy.
  - This is not visible in logs because print_args() is called at line 156 (before build_optimizer), so the printed warmup_bias_lr=0.1 does not reflect the actual value used
  during training.
  - Manually specifying optimizer=MuSGD avoids this issue and produces better results with the same lr and momentum.

  To Reproduce

  from ultralytics import YOLO

  model = YOLO("yolo11s-seg.pt")

  # Run 1: auto → MuSGD, warmup_bias_lr silently set to 0.0
  model.train(data="my_data.yaml", epochs=400, optimizer="auto")

  # Run 2: manual MuSGD, warmup_bias_lr stays at 0.1
  model.train(data="my_data.yaml", epochs=400, optimizer="MuSGD")

  # Run 2 produces noticeably higher accuracy than Run 1,
  # despite both using MuSGD with the same lr and momentum.

  Suggested Fix

  Move warmup_bias_lr = 0.0 inside the AdamW branch only:

  if iterations > 10000:
      name, lr, momentum = "MuSGD", 0.01, 0.9
  else:
      name, lr, momentum = "AdamW", lr_fit, 0.9
      self.args.warmup_bias_lr = 0.0  # no higher than 0.01 for Adam

  Environment

  - Ultralytics: 8.4.13
  - Python: 3.11
  - PyTorch: 2.9.1+cu128
  - GPU: NVIDIA A100-PCIE-40GB

  ---

### Environment

(YOLO) root@sx-dms0925:/data/lost+found/sjh/line_seg# yolo check
Ultralytics 8.4.13 🚀 Python-3.11.14 torch-2.9.1+cu128 CUDA:0 (NVIDIA A100-PCIE-40GB, 40444MiB)
Setup complete ✅ (96 CPUs, 754.4 GB RAM, 409.7/438.9 GB disk)

OS                     Linux-5.15.0-156-generic-x86_64-with-glibc2.35
Environment            Linux
Python                 3.11.14
Install                pip
Path                   /root/miniforge3/envs/YOLO/lib/python3.11/site-packages/ultralytics
RAM                    754.35 GB
Disk                   409.7/438.9 GB
CPU                    Intel Xeon Gold 6248R CPU @ 3.00GHz
CPU count              96
GPU                    NVIDIA A100-PCIE-40GB, 40444MiB
GPU count              8
CUDA                   12.8

numpy                  ✅ 2.2.0>=1.23.0
matplotlib             ✅ 3.10.7>=3.3.0
opencv-python          ✅ 4.12.0.88>=4.6.0
pillow                 ✅ 12.0.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.16.3>=1.4.1
torch                  ✅ 2.9.1>=1.8.0
torch                  ✅ 2.9.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.24.1>=0.9.0
psutil                 ✅ 7.1.3>=5.8.0
polars                 ✅ 1.35.2>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18

### Minimal Reproducible Example

```
from ultralytics import settings
settings["tensorboard"]=True
from ultralytics import YOLO

from pathlib import Path
import os
import torch
import gc
import sys
import argparse

# from ultralytics import RTDETR


def parse_args():
    parser = argparse.ArgumentParser(description="YOLO Training Script")
    parser.add_argument(
        "--exp_name",
        type=str,
        default=None,
        help="训练实验名称（保存目录名），如 yolo11x_text_v1"
    )
    return parser.parse_args()

if __name__ == '__main__':
    
    args = parse_args()

    # 加载模型
    weights = "./DATA/yolo26s-seg.pt"
    yaml = "ultralytics/cfg/models/26/yolo26-seg.yaml"
    model = YOLO(weights)
    # model = YOLO(yaml).load(weights)  # 从头开始构建新模型
    # model = YOLO("runs/detect/v5n9/weights/best.pt")  # 加载预训练模型（推荐用于训练）
    results = model.train(
        data="DATA/Complete_line/Complete_line_v1.0.yaml",
        epochs=400,
        batch=16,
        imgsz=640,
        workers=14,
        # name=args.exp_name or Path(model.cfg).stem,
        name=args.exp_name or Path(weights).stem, 
        multi_scale=0.25,
        save_period=100,
        # inner_iou=True,
        # inner_iou_version=2,
        # use_elongated_loss=True,
        # project="./runs",
        # device=[6]
        # lr0=0.005,
        optimizer= 'auto' # (str) optimizer: SGD, Adam, Adamax, AdamW, NAdam, RAdam, RMSProp, or auto
    )
```

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by knQzx**
fixes the issue where `optimizer=auto` unconditionally sets `warmup_bias_lr=0.0`, even when MuSGD is selected. this moves the assignment into the AdamW-only branch so MuSGD keeps its default `0.1`, matching the behavior of explicitly setting `optimizer=MuSGD`.

closes #23637

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Fixes `warmup_bias_lr` auto-configuration so it is only overridden for `AdamW`, not for `MuSGD` ⚙️

### 📊 Key Changes
- Updates `build_optimizer()` in `ultralytics/engine/trainer.py` to split the optimizer auto-selection into explicit `if/else` branches.
- Keeps `MuSGD` auto-selection unchanged for large training runs with `lr=0.01` and `momentum=0.9`.
- Restricts `self.args.warmup_bias_lr = 0.0` to the `AdamW` path only, instead of applying it to both optimizers.
- Preserves the existing Adam-specific safeguard comment indicating the bias warmup LR should remain low 🛠️

### 🎯 Purpose & Impact
- Prevents unintended mutation of `warmup_bias_lr` when `auto` selects `MuSGD` ✅
- Aligns optimizer-specific warmup behavior with the intended logic, reducing the chance of incorrect training hyperparameters.
- Improves training consistency and makes the auto-optimizer path easier to understand and maintain for contributors and users.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @knQzx, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help streamline review, and an Ultralytics engineer will assist you soon. Please review the checklist below to support a smooth integration of your fix for the optimizer warmup behavior ⚙️

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The change looks correct and appropriately scoped. Splitting the auto-optimizer selection into explicit branches fixes the unintended `warmup_bias_lr` mutation for `MuSGD` without altering the existing `AdamW` safeguard. I did not find any genuine issues in the shown diff.

**Comment by Y-T-G** [comment]
User is botting PRs

image
